rust/src/mqtt/parser.rs

```diff
@@ -222,99 +222,111 @@ pub fn parse_connect(i: &[u8]) -> IResult<&[u8], MQTTConnectData> {
     ))
 }
 
-pub fn parse_connack(i: &[u8], protocol_version: u8) -> IResult<&[u8], MQTTConnackData> {
-    let (i, topic_name_compression_response) = be_u8(i)?;
-    let (i, return_code) = be_u8(i)?;
-    let (i, properties) = parse_properties(i, protocol_version == 5)?;
-    Ok((
-        i,
-        MQTTConnackData {
-            session_present: (topic_name_compression_response & 1) != 0,
-            return_code,
-            properties,
-        },
-    ))
+#[inline]
+fn parse_connack(protocol_version: u8) -> impl Fn(&[u8]) -> IResult<&[u8], MQTTConnackData>
+where
+{
+    move |i: &[u8]| {
+        let (i, topic_name_compression_response) = be_u8(i)?;
+        let (i, return_code) = be_u8(i)?;
+        let (i, properties) = parse_properties(i, protocol_version == 5)?;
+        Ok((
+            i,
+            MQTTConnackData {
+                session_present: (topic_name_compression_response & 1) != 0,
+                return_code,
+                properties,
+            },
+        ))
+    }
 }
 
-pub fn parse_publish(
-    i: &[u8], protocol_version: u8, has_id: bool,
-) -> IResult<&[u8], MQTTPublishData> {
-    let (i, topic) = parse_mqtt_string(i)?;
-    let (i, message_id) = cond(has_id, be_u16)(i)?;
-    let (message, properties) = parse_properties(i, protocol_version == 5)?;
-    Ok((
-        i,
-        MQTTPublishData {
-            topic,
-            message_id,
-            message: message.to_vec(),
-            properties,
-        },
-    ))
+#[inline]
+fn parse_publish(protocol_version: u8, has_id: bool) -> impl Fn(&[u8]) -> IResult<&[u8], MQTTPublishData>
```

Comment:
I get it, but I find that functions that return functions often make things more complicate than they are.  Is there another way?

Reply:
I'm afraid that is the way it's done now if you want to have a parser combinator that takes more parameters than just the input, you need to produce a function that takes just the input. At least that's how I understood it.

Reply:
Thinking about this, I bet you can handle the `Incomplete` in only one place if you create the right intermediate function
`parse_message` calls some intermediate function to do `match message_type {`
If this intermediate function returns `Incomplete, `parse_message` transforms it into a plain error

What do you think ?

Reply:
True. See next PR.

Reply:
Looks good in next PR